keps/sig-node/5359-swap-awareness-api/README.md

```diff
@@ -0,0 +1,826 @@
+proposal adopts the position that limits control usage, not placement.
+
+-  The swap limits are managed at container level and placements are
+    determined at pod level. A "swap:0" container can be co-existing with
+    another workload utilizing swap.
+-  If workload separation for swap is desired, explicit placement controls
+    like taints or nodeSelector should be the preferred option, separating API
+    concerns of workload placement from resource usage.
+-  ‘limits' should not overload the meaning of "swap:0" to mean "I require a
+    non-swap node". Swap aware scheduling is investigated as a separate KEP
+    (xref: [#5424](https://github.com/kubernetes/enhancements/issues/5424)). 
+
+### Risks and Mitigations
+
+<<[UNRESOLVED kannon@]>>
+
+1. **Risk: Discuss the implications of overcommitting swap further**
+
+-  what k8s should do to make sure node doesn't end up in a place where all
+    the pods have swap provisioned but cannot utilize anymore.
+
+> This can be addressed by the user responding with configuring an
+additional swap. K8s cannot react to swap is full as swap is a node
+resource; with better observability story this concerns could be reduced.  
+  
+
+-  better observability story; user should be able to know when there is
+    overcommit of swap or there will be swap capacity crunch.
+
+> We already have swap metrics for capacity and usage at node, pod and
+container level
+> - Would a new metric for `kubelet_node_swap_allocated_bytes `address this
+concern?
+> - This would be sum of all `resources.limits.swap` for all containers
+running on that node. This can help operators to create precise alert for their
+risk for ( allocated / capacity ).  
+> - New Condition for `SwapPressure` in NPD.
+
+<<[/UNRESOLVED]>>
+
+1. Risk: User confusion between `LimitedSwap` and `WorkloadControlledSwap`
+    modes.
+
+Mitigation: Swap behavior will be exposed as a field in node-info to be
+observable by the user.
+## Design Details
+
+### Node Configuration
+
+A new `swapBehavior` is introduced in the `kubeletConfiguration`
+
+```
+kubeletConfiguration:
+  memorySwap:
+    swapBehavior: "WorkloadControlledSwap" # Node-level swap enabled, but workloads control usage
+```
+### Proposed Design: Limits-Only Model
+
+Swap limits are configured per container for a cleaner resource model. This
+avoids the ambiguity of swap requests.
+
+-  **Rationale:** "policy" fits per pod, swap "limits" are container specific
+    as swap is treated by kernel per process. Starting with ‘container' limits
+    first gives us flexibility for unambiguous design. If we start with pod
+    limits first, this implies all containers and we will need to reconsider
+    how to support individual container limits in the future. (eg: will it
+    override?)
+-  This also avoids handling conflicts with current `PodLevelResource`
+    behavior of applying limit as request and using for admission time.   
+
+```yaml
+resources:
+  limits:
+    memory: "2Gi"
```

Comment:
One downside to this approach BTW is the lack of ability to express simple logic, e.g. "I want 10% of swap as a limit, but not less than 500Mi". One way to solve this problem with this API is to set it via an external controller post-scheduling.

I wonder if, moving forward, we should come up with something like an admin-level CRD that will allow matching swap configurations to workloads.